tests/unit/model/relationships/has-many-test.js

```diff
@@ -906,6 +906,561 @@ test('it is possible to add a new item to a relationship', function(assert) {
+  assert.equal(get(pets, 'length'), 1, 'precond - relationship has only one pet to start');
+
+  run(() => {
+    pets.pushObjects([rambo, rebel]);
+  });
+
+  assert.equal(get(pets, 'length'), 3, 'precond2 - relationship now has three pets');
+
+  run(() => {
+    return shen.destroyRecord({})
+      .then(() => {
+        shen.unloadRecord();
+      });
+  });
+
+  assert.equal(get(pets, 'length'), 2, 'relationship now has two pets');
+});
+
+test('new items added to an async hasMany relationship are not cleared by a delete', function(assert) {
+  assert.expect(7);
+
+  const Person = DS.Model.extend({
+    name: DS.attr('string'),
+    pets: DS.hasMany('pet', { async: true, inverse: null })
+  });
+
+  const Pet = DS.Model.extend({
+    name: DS.attr('string'),
+    person: DS.belongsTo('person', { async: false, inverse: null })
+  });
+
+  let env = setupStore({
+    person: Person,
+    pet: Pet
+  });
+  env.adapter.shouldBackgroundReloadRecord = () => false;
+  env.adapter.deleteRecord = () => {
+    return Ember.RSVP.Promise.resolve({ data: null });
+  };
+
+  let { store } = env;
+
+  run(() => {
+    store.push({
+      data: {
+        type: 'person',
+        id: '1',
+        attributes: {
+          name: 'Chris Thoburn'
+        },
+        relationships: {
+          pets: {
+            data: [
+              { type: 'pet', id: '1' }
+            ]
+          }
+        }
+      },
+      included: [
+        {
+          type: 'pet',
+          id: '1',
+          attributes: {
+            name: 'Shenanigans'
+          }
+        },
+        {
+          type: 'pet',
+          id: '2',
+          attributes: {
+            name: 'Rambunctious'
+          }
+        },
+        {
+          type: 'pet',
+          id: '3',
+          attributes: {
+            name: 'Rebel'
+          }
+        }
+      ]
+    });
+  });
+
+  return run(() => {
+    const person = store.peekRecord('person', '1');
+    const petsProxy = run(() => person.get('pets'));
+
+    return petsProxy.then((pets) => {
+      const shen = pets.objectAt(0);
+      const rambo = store.peekRecord('pet', '2');
+      const rebel = store.peekRecord('pet', '3');
+
+      assert.equal(get(shen, 'name'), 'Shenanigans', 'precond - relationships work');
+      assert.equal(get(pets, 'length'), 1, 'precond - relationship has only one pet to start');
+      assert.equal(get(petsProxy, 'length'), 1, 'precond - proxy has only one pet to start');
+
+      pets.pushObjects([rambo, rebel]);
+
+      assert.equal(get(pets, 'length'), 3, 'precond2 - relationship now has three pets');
+      assert.equal(get(petsProxy, 'length'), 3, 'precond2 - proxy now reflects three pets');
+
+      return shen.destroyRecord({})
+        .then(() => {
+          shen.unloadRecord();
+
+          assert.equal(get(pets, 'length'), 2, 'relationship now has two pets');
+          assert.equal(get(petsProxy, 'length'), 2, 'proxy now reflects two pets');
```

Comment:
let's change these assertions to `deepEqual` an array of ids rather than just checking the length